### ISSUE TITLE
SerialPort API Update

### ISSUE THREAD

**Issue by BenjaminHolland** [dotnet/corefx#33560]
The API for the serial port has barely changed since it was introduced, despite significant changes in language features and even tools. Now that the package can be updated OOB, it would seem like a good time to finally update this component, or do something similar to what happened with the .net JSON serialize and designate a pseudo-official third party option. I'd like to start brainstorming how this API should evolve. 

### Async
Applications utilizing serial ports are arguably more susceptible to slowdowns than network apps due to the relatively slow speeds of serial ports. Having an API that forces an app blocking until a device that's communicating at 2800 baud completes the transfer when every other IO facility, most of which are consistently faster, has had some form of asynchronous ability built in is just sad.

I propose adding async overloads for then Read, Write, Open, and Close methods. For targets that don't support async, Begin/End methods should be provided instead. These methods should respect the timeout settings on the port, o r provide additional overloads that specify a timeout. 

### Memory/Span Overloads
The work being done on efficient buffers is extremely useful for applications using serial ports. Serial Ports are often used in IoT and embedded applications where memory and processing power is at a premium. 

I propose adding Read/Write overloads that use Span<byte/char> and ReadOnlySpan<byte/char>. This would allow easy integration with the Pipelines namespace. 

### Stream Compatibility
It would be extremely useful to be able to treat the serial port as a stream, but using the serial ports BaseStream property is more or less undocumented and it's behavior is undefined. Something should be done to make this compatible with core IO stream utilities.
Options:

- Provide a method for producing a reader and a writer stream
- Provide documentation for the BaseStream
- Change SerialPort to directly subclass BaseStream
- Attach the serial port interface directly to the SerialStream class and expose it.

### Mocking/Testing Interface
Not having an interface for a serial port complicates testing. Providing an interface would make this easier. 

### Remove or Explain Component Superclass.
I'm not sure why, and I can't find any justification for, attaching UI dev facilities directly onto a mid-level IO facility. Socket isn't a component. Nor are Pipes. Both seem to allow remoting by inheriting from `MarshalByRefObject`. Can this dependency be removed? Especially in .net core, it seems out of place.

**Comment by shaggygi** [comment]
Referencing...
- https://github.com/dotnet/corefx/issues/18012
- https://github.com/dotnet/corefx/issues/33374

**Comment by krwq** [comment]
So just for reference SerialPort.BaseStream has async APIs and AFAIK they currently work on Linux - I saw Windows is using completion ports so I presume that also works but I haven't tested.

**Comment by BenjaminHolland** [comment]
@krwq Yeah, BaseStream "works" in the sense that you can call ReadAsync and WriteAsync and have them work under ideal conditions. I can only assume they follow the behavior described in the `Stream` documentation.

**Comment by hansmbakker** [comment]
What about an implementation of [`IDuplexPipe`](https://docs.microsoft.com/en-us/dotnet/api/system.io.pipelines.iduplexpipe?view=dotnet-plat-ext-2.1) for SerialPort?

**Comment by BenjaminHolland** [comment]
That seems like a great idea, but I haven't looked at the Pipelines project
in a bit. From what i remember it supports most of the operations.
